### PR TITLE
UTY-329: Automatic disconnect at the start of compilation.

### PR DESCRIPTION
**Pull request by martinzlocha**
**Contributions**: We are not currently taking public contributions - see our [contributions](https://github.com/spatialos/UnityGDK/blob/master/CONTRIBUTING.md) policy. However, we are accepting issues and we do want your [feedback](https://github.com/spatialos/UnityGDK/blob/master/README.md#give-us-feedback).

-------

#### Description
Editor hangs frequently if code compiles while the editor is playing and connected to Spatial.

**Do not merge:** This fixes the issue of the editor freezing there are still issues caused by lack of cleanup after disconnect. This is blocked by UTY-339.

#### Tests
Connected to a local deployment, changed some code, recompiled. The editor doesn't hang.

#### Documentation
N/A

#### Primary reviewers
If your change will take a long time to review, you can name at most two primary reviewers who are ultimately responsible for reviewing this request. @ mention them.
@JonasImprobable @zeroZshadow @jessicafalk

### CODE REVIEW

File: workers/unity/Assets/Playground/Scripts/Player/ProcessLaunchCommandSystem.cs [hunk 1]
```diff
@@ -18,8 +18,7 @@ private struct LaunchCommandData
             [ReadOnly] public EntityArray Entity;
             public ComponentDataArray<SpatialOSLauncher> Launcher;
 
-            [ReadOnly]
-            public ComponentArray<CommandRequests<Generated.Playground.Launcher.LaunchEntity.Request>> CommandRequests;
+            [ReadOnly] public ComponentArray<CommandRequests<Generated.Playground.Launcher.LaunchEntity.Request>> CommandRequests;
```
Comment (jamiebrynes7) [inline workers/unity/Assets/Playground/Scripts/Player/ProcessLaunchCommandSystem.cs]: I've had linting fail this yesterday in the cloud FYI (although it doesn't seem like it does all the time)

Comment (martinzlocha) [reply workers/unity/Assets/Playground/Scripts/Player/ProcessLaunchCommandSystem.cs]: I mean these are the changes made when I ran ci/lint.sh so it shouldn't fail. *He says wishfully*

File: workers/unity/Assets/Gdk/Core/Worker/WorkerBase.cs [hunk 2]
```diff
@@ -61,6 +61,18 @@ public bool Connect(ConnectionConfig config)
...
         }
 
+        public void Disconnect()
+        {
+            if (Connection == null)
+            {
+                return;
+            }
+
+            View.Disconnect("WorkerBase.Disconnect() called.");
```
Comment (jared-improbable) [inline workers/unity/Assets/Gdk/Core/Worker/WorkerBase.cs]: Is this a meaningful disconnection reason for users? Can it be made less specific? The hardcoded `WorkerBase.Disconnect()` is a potential source of problems if this is refactored/renamed at some point.

File: workers/unity/Assets/Playground/Scripts/Bootstrap.cs [hunk 3]
```diff
@@ -90,6 +92,14 @@ public void Awake()
             World.Active = worlds[0];
         }
 
+        private static void CompilationStarted(String destination)
```
Comment (jared-improbable) [inline workers/unity/Assets/Playground/Scripts/Bootstrap.cs]: Use `string` instead.

File: workers/unity/Assets/Playground/Scripts/Bootstrap.cs [hunk 4]
```diff
@@ -28,6 +29,7 @@ public void Awake()
             SetupInjectionHooks(); // Register hybrid injection hooks
             PlayerLoopManager.RegisterDomainUnload(DomainUnloadShutdown, 10000); // Clean up worlds and player loop
 
+            CompilationPipeline.assemblyCompilationStarted += CompilationStarted;
```
Comment (jared-improbable) [inline workers/unity/Assets/Playground/Scripts/Bootstrap.cs]: Should this also be removed from `assemblyCompilationStarted`?

File: workers/unity/Assets/Playground/Scripts/Player/ProcessLaunchCommandSystem.cs (hunk 1 above)
Comment (jamiebrynes7) [reply workers/unity/Assets/Playground/Scripts/Player/ProcessLaunchCommandSystem.cs]: Yeah its not consistent at all, its really sad

File: workers/unity/Assets/Gdk/Core/Worker/WorkerBase.cs (hunk 2 above)
Comment (martinzlocha) [reply workers/unity/Assets/Gdk/Core/Worker/WorkerBase.cs]: Yup, changed :)

File: workers/unity/Assets/Playground/Scripts/Bootstrap.cs (hunk 4 above)
Comment (martinzlocha) [reply workers/unity/Assets/Playground/Scripts/Bootstrap.cs]: Added removal as well :)

Comment (zeroZshadow) [reply workers/unity/Assets/Playground/Scripts/Bootstrap.cs]: This is unity editor only, and will fail compilation for standalone!

Comment (zeroZshadow) [reply workers/unity/Assets/Playground/Scripts/Bootstrap.cs]: Please move this into the `UNITY_EDITOR` block

Comment (martinzlocha) [reply workers/unity/Assets/Playground/Scripts/Bootstrap.cs]: Ahh thanks for that, moved.

**Comment by martinzlocha** [comment]
FYI. I realized that even though this fixes the issue of the editor freezing there are still issues caused by lack of cleanup after disconnect. That has to be merged in first.

File: workers/unity/Assets/Playground/Scripts/Bootstrap.cs [hunk 5]
```diff
@@ -115,7 +116,7 @@ public static void InitializeWorkerTypes()
             WorkerRegistry.RegisterWorkerType<UnityGameLogic>();
         }
 
-        public static void SetupInjectionHooks()
+        public static void SetupInjectionHooks()
```
Comment (zeroZshadow) [inline workers/unity/Assets/Playground/Scripts/Bootstrap.cs]: Whitespace?

File: workers/unity/Assets/Gdk/Core/Worker/WorkerBase.cs [hunk 6]
```diff
@@ -61,6 +61,18 @@ public bool Connect(ConnectionConfig config)
             return true;
         }
 
+        public void Disconnect()
+        {
+            if (Connection == null)
+            {
+                return;
```
Comment (samiwh) [inline workers/unity/Assets/Gdk/Core/Worker/WorkerBase.cs]: This probably shouldn't do nothing. Until we start looking at the broader error stuff I'd log this. I don't think this should do nothing though.

File: workers/unity/Assets/Gdk/Core/Worker/WorkerBase.cs [hunk 7]
```diff
@@ -61,6 +61,18 @@ public bool Connect(ConnectionConfig config)
...
+            }
+
+            View.Disconnect("Disconnect called.");
```
Comment (samiwh) [inline workers/unity/Assets/Gdk/Core/Worker/WorkerBase.cs]: Please change this to something which makes it clear that it was disconnected internally. As a user seeing this while debugging will be confusing.

Comment (martinzlocha) [reply workers/unity/Assets/Gdk/Core/Worker/WorkerBase.cs]: Updated, should hopefully be better,

File: workers/unity/Assets/Gdk/Core/Worker/WorkerBase.cs [hunk 8]
```diff
@@ -66,6 +66,19 @@ public bool Connect(ConnectionConfig config)
             return true;
         }
 
+        public void Disconnect()
+        {
+            if (Connection == null)
+            {
+                Debug.LogError("Attempted to disconnect but connection is already null.");
```
Comment (jamiebrynes7) [inline workers/unity/Assets/Gdk/Core/Worker/WorkerBase.cs]: Maybe `Attempted to disconnect, but {WorkerId} is not connected`?

**Comment by martinzlocha** [comment]
Yup after discussion with @jessicafalk it would take too long to merge because its blocked. Will close for now.

**Pull request closed without merging**